[90X] Allow multi-IOV input for Millepede alignment framework

ghellwig opened the pull request:
Removes shortcoming of the alignment framework which has not been able to update the alignables after an update of the tracker geometry at an IOV boundary.

Removes also temporary fix of #16135 and enables the usage of multi-IOV input for Millepede.

Added various protection against misconfiguration.
 - consistency check of the configuration during the setup of the alignment campaign
 - consistency checks in the C++ code to detect misconfiguration

The code has been used to derive the 2016 legacy alignment.

The general AlignmentProducer has the new features disabled by default. It is only enabled for Millepede. If other alignment algorithms want to use it a flag has to be set to `True`:
```
process.AlignmentProducer.enableAlignableUpdates = True
```

The code is based on top of #17408, i.e. to better digest the changes, it might be useful to first review and eventually approve #17408.

EDIT: A dedicated presentation on this update has been given at the [Tracker Alignment meeting on 01-Mar-2017](https://indico.cern.ch/event/607508/contributions/2493829/attachments/1420329/2176319/Alignment-MultiIOV_CMS-TkAl_20170301.pdf)

gregor-mittag committed:
Fix 'mps_parse_pedechi2hist.py'.

gregor-mittag committed:
Fix variable replacement in pede-script-template.

gregor-mittag committed:
Fix style of 'mps_check.py'.

gregor-mittag committed:
Add handling of missing output file (rare case).

gregor-mittag committed:
Use context manager to open files in 'mps_check.py'.

gregor-mittag committed:
Add protection against accidentally using data prior to first IOV.

gregor-mittag committed:
Ignore auto-generated ROOT file in MPS validation.

gregor-mittag committed:
Fix style of 'mps_fetch.py'.

gregor-mittag committed:
Add recovery from previous failures in 'mps_fetch.py'.

gregor-mittag committed:
Partly align AlignmentProducer to PCLTrackerAlProducer.

 - move subtask into dedicated methods
 -> first step needed for unification
 -> also needed for support of multi-IOV input tag

gregor-mittag committed:
Add check if setup has changed to 'AlignmentProducer'.

gregor-mittag committed:
Fixed memory usage calculation.

gregor-mittag committed:
Fix input validation in 'mps_create_file_lists.py'.

gregor-mittag committed:
Fix automatic single-IOv input for '-w' option of 'mps_alisetup.py'.

gregor-mittag committed:
Added SurfaceDeformations to 'AlignableDetUnit::dump()'.

gregor-mittag committed:
Added check for pede step, if all needed runs have been processed.

gregor-mittag committed:
Added flag to enable/disable updatable alignables.

gregor-mittag committed:
Move common function to utilities.

gregor-mittag committed:
Added 'update' method to 'Alignable'.

gregor-mittag committed:
Added 'update' method to 'AlignableDetUnit'.

gregor-mittag committed:
Added 'reset' method to 'AlignableBeamSpot'.

gregor-mittag committed:
Added 'update' methods to 'AlignableComposite'.

gregor-mittag committed:
Added 'update' method to 'AlignableDet'.

gregor-mittag committed:
Added 'update' methods to muon AlignableDets.

gregor-mittag committed:
Added beam-spot resetting function to 'AlignableExtras'.

gregor-mittag committed:
Update alignables at IOV boundaries in AligmmentProducer.

gregor-mittag committed:
Add run-based caching to Alignables.

gregor-mittag committed:
Add option to skip check for GlobalPositionRcd IOV boundary.

gregor-mittag committed:
Removed temporary protection against multi-IOV input.

 - still enforce it in case no output IOVs are specified
 - adapt MPS to handle multi-IOV inputs properly

gregor-mittag committed:
Support updates in AlignableTrackerBuilder and AlignableCompositeBuilder.

gregor-mittag committed:
Added 'update' methods to muon 'AlignableTracker' and 'AlignableMuon'.

gregor-mittag committed:
Further improve input validation of mps_create_file_lists.py.

gregor-mittag committed:
Make mps_create_file_lists.py more robust against DAS issues.

gregor-mittag committed:
Fixed shuffling of files in case a DAS cache is used.

gregor-mittag committed:
Create dataset cff-files also for alignment datasets.

gregor-mittag committed:
Enable forwarding of proxy for MPS.

gregor-mittag committed:
Fix '-f/--force-merge' option of 'mps_fire.py'.

gregor-mittag committed:
Allow multiple levels of IOVs.

gregor-mittag committed:
Fix behaviour of mps_create_file_lists.py if only help is wanted.

gregor-mittag committed:
Speed-up of file creation script.

When requesting information of all 2016 MinBias datasets, the runtime is reduced
from 2-3 hours to approx. 2 minutes.

gregor-mittag committed:
Better log message for file-list creation script.

gregor-mittag committed:
Enable usage of UpsilonMuMu data in MillePede workflow.

Changes commmited on behalf of Rainer Mankel.
Details are presented here:
https://indico.cern.ch/event/607492/#35-update-on-upsilonmumu-const

gregor-mittag committed:
Define default pede settings and provide helper to customize them.

gregor-mittag committed:
Adapt single-IOV creation script to sqlalchemy 1.1.4

gregor-mittag committed:
Move common function to mps tools.

gregor-mittag committed:
Separate process extraction from input-db creation.

gregor-mittag committed:
Implemented early detection of inconsistent IOV definitions in MPS.

 - split up input db creation in separate functions
 - propagate this change to stand-alone input creation script

gregor-mittag committed:
Added unit test for database-query script.

gregor-mittag committed:
Allow multi-IOV input in pixel-barycenter tool.

gregor-mittag committed:
Remove redundant headers.

ghellwig commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @ghellwig (Gregor Mittag) for CMSSW_9_0_X.

It involves the following packages:

Alignment/CommonAlignment
Alignment/CommonAlignmentAlgorithm
Alignment/CommonAlignmentProducer
Alignment/MillePedeAlignmentAlgorithm
Alignment/MuonAlignment
Alignment/TrackerAlignment


@ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @mmusich, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @pakhotin, @abbiendi, @jhgoh, @tocheng, @tlampen, @mschrode, @mmusich, @trocino this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

cmsbuild commented:
-1

Tested at: 831c2e9d9d8a266b013b2d10c6ae899be0c381de

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/17716/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1001.0 step3
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step3_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

mmusich commented:
hi Gregor, @ghellwig 
```
----- Begin Fatal Exception 09-Feb-2017 17:40:40 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PCLTrackerAlProducer label='SiPixelAliMilleAlignmentProducer'
Exception Message:
MissingParameter: The required parameter 'enableAlignableUpdates' was not specified.
----- End Fatal Exception -------------------------------------------------
```
looks like the new parameter was somehow missed in the configuration here:

https://github.com/ghellwig/cmssw/blob/831c2e9d9d8a266b013b2d10c6ae899be0c381de/Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAli_cff.py#L76-L106

ghellwig commented:
Hi Marco,
yes, I had to leave in a hurry yesterday and overlooked that I did not commit this change.
It's time to finalize the unification of the separate PCL and offline alignment producers...
Fortunately, this PR brings us also a bit closer to this goal, i.e. less duplication of code and configurations.

I'll push it later because I currently have problems at lxplus.

gregor-mittag committed:
Adjust PCL alignment producer to multi-IOV alignment framework.

 - no functional change
 - just added missing configuration parameters which are deactivated for PCL

cmsbuild commented:
Pull request #17475 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @mmusich, @davidlange6 can you please check and sign again.

ghellwig commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: edfcd6466a1ed714655096dcf3ca561e4fb812bf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/17732/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/17732/summary.html

ghellwig commented:
@smuzaffar I just checked the clang results and found that a unit test within this PR does not pass there.
It happens because the script used for this unit test is taken from the release.
This PR fixes this script and consequently the test passes in the unit tests section, but it would be nice if the clang check could acknowledge possible local changes, too.

smuzaffar commented:
@ghellwig , I noticed that new ```<test name=... command=.../>``` is also being run at the build time (which is a bug). I have a fix for that which I am testing right now. So please ignore the clang error in this PR. The unit test testPrepareInputDb actually ran fine when it was run after the build.

ghellwig commented:
@smuzaffar I am fine with ignoring it, but just wanted to report it :)

gregor-mittag committed:
Fix clang warnings caused by 'update' method.

Make the method non-virtual because it acts like a constructor applied
on an already constructed object.
Since a constructor is always non-virtual it appears to be cleanest to
declare also the 'update' methods non-virtual.

cmsbuild commented:
Pull request #17475 was updated. @ghellwig, @arunhep, @cerminar, @cmsbuild, @franzoni, @mmusich, @davidlange6 can you please check and sign again.

ghellwig commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ffa383dedd31d38ddadebbf75dbf118b07264dd1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/17775/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/17775/summary.html

ghellwig commented:
@smuzaffar @davidlange6 I have a naive question:
Shouldn't the list of added commits in this PR be updated after #17408 got merged? Or at least the diff should not show the changes introduced by #17408 anymore, right?

PS: This PR was based on #17408, and I thought github should be clever enough to detect that. The check-mark at https://github.com/cms-sw/cmssw/pull/17475/commits/3815b8d2d7044806b35f4ab0a1e046ff344c4b56 above also indicates that my assumption is not completely wrong.

smuzaffar commented:
@ghellwig , ideally yes but no idea why github does not recognize this. I can check with the GH support team

ghellwig commented:
@franzoni to ease the digestion of this PR:
A dedicated presentation on this update has been given at the [Tracker Alignment meeting on 01-Mar-2017](https://indico.cern.ch/event/607508/contributions/2493829/attachments/1420329/2176319/Alignment-MultiIOV_CMS-TkAl_20170301.pdf).

franzoni commented:
pelase test
thanks @ghellwig

ghellwig commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ffa383dedd31d38ddadebbf75dbf118b07264dd1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/18202/summary.html

cmsbuild commented:
Comparison job queued.

franzoni commented on Alignment/MillePedeAlignmentAlgorithm/python/mpslib/Mpslibclass.py:
```diff
@@ -99,18 +99,18 @@ def read_db(self, db_file_name = __default_db_file_name):
+            self.JOBREMARK.append(parts[9].strip())
+            self.JOBSP1.append(parts[10].strip())
+            self.JOBSP2.append(parts[11].strip())
+            self.JOBSP3.append(parts[12].strip())
```
hello @ghellwig 
how do you know which elements of parts[]
you're in need of removing blanks from ?

ghellwig replied:
Everything that is not converting to an integer remains a string and can have unintended whitespace at the left and right end.
The conversion to integers ignores such whitespace anyways.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17475/18202/summary.html

franzoni commented:
no changes in plots 2017 ttbar, 2016B data, runII2016 25 and 50 ns

franzoni commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

franzoni commented:
including this PR 900 important for alignment 
may we have this PR merged, please ?

davidlange6 commented:
+1
